### DISCUSSION TITLE
Different behaviour promised vs. awaited Element

### DISCUSSION THREAD

**Discussion by synedra-mpe** [webdriverio/webdriverio#8744] [Technical Discussion]
Is it expected beahaviour or a bug, that there is a difference in the two usages of `browser.waitUntil()`? 
```javascript
const foo = await $('#username');
const bar = $('#username');
await browser.waitUntil(foo.isExisting); // does not work - times out
await browser.waitUntil(bar.isExisting); // works
```

**Comment by christian-bromann** [comment]
WebdriverIO provides `browser.waitForExisting()` for that purpose. I assume the reason why `await browser.waitUntil(foo.isExisting)` times out is because the scope is incorrect. I guess `await browser.waitUntil(foo.isExisting.bind(foo))` will work. The latter works because the execution is chained and the scope is being maintained.

**Comment by erwinheitzman** [comment]
Wait wait, the actual issue here is this:
```
const foo = await $('#username');
const bar = $('#username');

await browser.waitUntil(() => foo.isExisting()); // works
await browser.waitUntil(() => bar.isExisting()); // works

await browser.waitUntil(foo.isExisting); // does not work - times out
await browser.waitUntil(bar.isExisting); // works
```

i do not understand why the third example is an exception to the rule here and asked @synedra-mpe to share this here so that other team members can also take a look :)

**Reply by erwinheitzman** [reply]
I tried the `await browser.waitUntil(foo.isExisting.bind(foo))` and that indeed solves the problem @christian-bromann is this something we should address since I think this can happen for any command right?

**Reply by christian-bromann** [reply]
the difference between `foo` and `bar` is that `bar` is a chained element where there is some JS magic to allow continue call commands on it. Within the chaining we contain the scope which is why the example works for this case. Calling `await browser.waitUntil(foo.isExisting)` will loose the scope hence, you have to bind the scope to it.

> is this something we should address since I think this can happen for any command right?

I don't think we need to address anything here given that we should not recommend to wait on an element like this. We have build `waitForExists` for that reason and it has the same expressiveness.
